### PR TITLE
[Syntax] Add syntax support for `@differentiating` attribute.

### PR DESCRIPTION
**Pull request by dan-zheng**
Add syntax support for `@differentiating` attribute.
- Refactor `@differentiable` attribute syntax support to share syntax nodes.
- Pre-emptively add syntax support for `wrt` parameter list.

Resolves [TF-362](https://bugs.swift.org/browse/TF-362).
Up next: supporting generic constraints and `wrt` parameter list for `@differentiating` attribute.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: utils/gyb_syntax_support/AttributeNodes.py [hunk 1]
```diff
@@ -267,19 +263,49 @@
...
-             Child('DeclBaseName', kind='Syntax', description='''
+             Child('FunctionDeclName', kind='FunctionDeclName',
+                   description='The referenced function name.'),
+             Child('TrailingComma', kind='CommaToken', is_optional=True),
+         ]),
+
+    # func-decl-name -> (identifier | operator) decl-name-arguments?
+    # NOTE: This is duplicated with `DeclName` above. Change `DeclName`
+    # description and use it if possible.
+    Node('FunctionDeclName', kind='Syntax',
```
Comment (dan-zheng) [inline utils/gyb_syntax_support/AttributeNodes.py]: `FunctionDeclName` is duplicated with the [`DeclName`](https://github.com/apple/swift/blob/706a75007ba6d6569b248210cbf7e900c3df1b03/utils/gyb_syntax_support/AttributeNodes.py#L126) syntax node.
(The extra `SpacedBinaryOperator` node choice is added to support parsing `-`, which gets parsed as a `SpacedBinaryOperator` and not a `PrefixOperator`).

The reservations for not using `DeclName` are:
* The descriptions for `DeclName` are too specific and need to be made more general. Currently, they're specialized for protocol requirements, e.g. `"The base name of the protocol's requirement."`
* We need to add a `SpacedBinaryOperator` node choice.

@nkcsgexi: is it okay to edit `DeclName` (change descriptions and add node choice), or is that breaking?

**Review by rxwei — APPROVED** [review]
Great! This should unblock https://github.com/apple/swift/pull/23193.

**Comment by dan-zheng** [comment]
Unfortunately, I tried cherry-picking #23193 but it crashes:
```
Assertion failed: (Storage.hasVal), function getPointer, file /Users/danielzheng/swift-build/llvm/include/llvm/ADT/Optional.h, line 176.
Stack dump:
...
2.	While running pass #114915 SILModuleTransform "Differentiation".
0  swift                    0x0000000110c106e8 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift                    0x0000000110c0f6b8 llvm::sys::RunSignalHandlers() + 248
2  swift                    0x0000000110c10d02 SignalHandler(int) + 258
3  libsystem_platform.dylib 0x00007fff68c89f5a _sigtramp + 26
4  libsystem_platform.dylib 000000000000000000 _sigtramp + 2536988864
5  libsystem_c.dylib        0x00007fff68a271ae abort + 127
6  libsystem_c.dylib        0x00007fff689ef1ac basename_r + 0
7  swift                    0x000000010db13278 swift::SILFunctionType::getAutoDiffAssociatedFunctionType(llvm::SmallBitVector const&, unsigned int, unsigned int, swift::AutoDiffAssociatedFunctionKind, swift::SILModule&, llvm::function_ref<llvm::Optional<swift::ProtocolConformanceRef> (swift::CanType, swift::Type, swift::ProtocolDecl*)>, swift::GenericSignature*) + 4216
8  swift                    0x000000010d77552e (anonymous namespace)::DifferentiationTask::DifferentiationTask((anonymous namespace)::ADContext&, swift::SILFunction*, swift::SILDifferentiableAttr*&&, (anonymous namespace)::DifferentiationInvoker) + 1646
9  swift                    0x000000010d76f914 (anonymous namespace)::ADContext::registerDifferentiationTask(swift::SILFunction*, swift::SILAutoDiffIndices const&, (anonymous namespace)::DifferentiationInvoker) + 388
10 swift                    0x000000010d76cf04 (anonymous namespace)::Differentiation::run() + 1252
11 swift                    0x000000010d86ef3b swift::SILPassManager::runModulePass(unsigned int) + 1083
12 swift                    0x000000010d86fa14 swift::SILPassManager::execute() + 692
13 swift                    0x000000010cd45eab swift::SILPassManager::executePassPipelinePlan(swift::SILPassPipelinePlan const&) + 187
14 swift                    0x000000010d8788c4 swift::runSILDiagnosticPasses(swift::SILModule&) + 132
15 swift                    0x000000010cbf194b performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 11291
16 swift                    0x000000010cbedc9d swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3021
17 swift                    0x000000010cb9f88e main + 686
18 libdyld.dylib            0x00007fff6897b015 start + 1
```

[Full log here.](https://gist.github.com/dan-zheng/73ec39b669c9a38b43c7d5c507b431c0) I'll investigate soon.

File: utils/gyb_syntax_support/AttributeNodes.py [hunk 2]
```diff
@@ -267,19 +263,49 @@
...
                    The argument labels of the referenced function, optionally \
                    specified.
                    '''),
-             Child('TrailingComma', kind='CommaToken', is_optional=True),
+         ]),
+
+    # SWIFT_ENABLE_TENSORFLOW
+    # The argument of '@differentiating(...)'.
+    # differentiating-attr-arguments ->
+    #     differentiable-attr-func-specifier # original
```
Comment (dan-zheng) [inline utils/gyb_syntax_support/AttributeNodes.py]: This comment is incorrect:
```suggestion
    #     func-decl-name
```
I'll update it in my next PR.

File: utils/gyb_syntax_support/AttributeNodes.py (hunk 1 above)
Comment (dan-zheng) [reply utils/gyb_syntax_support/AttributeNodes.py]: I'll proceed with the duplicated `FunctionDeclName` syntax node for now.

File: utils/gyb_syntax_support/AttributeNodes.py (hunk 2 above)
Comment (dan-zheng) [reply utils/gyb_syntax_support/AttributeNodes.py]: Fixed in https://github.com/apple/swift/pull/23350/files.

**Pull request merged**